server/models.py

```diff
@@ -1643,3 +1644,93 @@ class CanvasAssignment(Model):
     @property
     def url(self):
         return '{}/assignments/{}'.format(self.canvas_course.url, self.external_id)
+
+#########
+# Files #
+#########
+
+class ExternalFile(Model):
+
+    id = db.Column(db.Integer, primary_key=True)
+    # Bucket/Folder that the file is stored in
+    container = db.Column(db.String(255), nullable=False)
+    filename = db.Column(db.String(1024), nullable=False)
+    object_name = db.Column(db.String(1024), nullable=False)
+
+    is_staff = db.Column(db.Boolean, nullable=False)
+    deleted = db.Column(db.Boolean, default=False)
+
+    course_id = db.Column(
+        db.Integer, db.ForeignKey('course.id'), index=True, nullable=False
+    )
+    course = db.relationship('Course')
+
+    user_id = db.Column(
+        db.Integer, db.ForeignKey('user.id'), index=True, nullable=False
+    )
+    user = db.relationship('User')
+
+    assignment_id = db.Column(
+        db.Integer, db.ForeignKey('assignment.id'), index=True
+    )
+    assignment = db.relationship('Assignment')
+
+    @property
+    def object(self):
+        return storage.get_blob(self.object_name, self.container)
+
+    @property
+    def mimetype(self):
+        guess = mimetypes.guess_type(self.filename)
+        if not guess[0]:
+            return 'application/octet-stream'
+        return guess[0]
+
+    def delete(self):
+        self.object.delete()
+        self.deleted = True
+        db.session.commit()
+
+    @staticmethod
+    def upload(iterable, user_id, name, is_staff=True, course_id=None,
```

Comment:
I'm going to move this over into Storage. I initially put it here to avoid circular dependencies - but that never happened - it makes more sense in `storage.py`